### ISSUE TITLE
docker harvester stops if recursive dir is empty or errors at some point.

### ISSUE THREAD

**Issue by chilio** [Chia-Network/chia-docker#187]
This is kind of a feature request, but I think it is pretty important.
When this image is run with `-e recursive_plot_scan="true" `  and container hits empty or errored target directory/mount, this harvester becomes obsolete with zero plots.
I believe the correct behavior would be to read all dirs (under the selected parent directory) and add all available/accesible plots to the chia harvester, even though some dirs might not be readable.

**Comment by Starttoaster** [comment]
For my context, is that the behavior of the Chia harvester process, or a behavior you've noticed to be specific to the harvester when ran inside a docker container?

**Comment by chilio** [comment]
I though the `-e recursive_plot_scan="true"` directive works only in docker environment, so I haven't tested this any other way.
And I can double confirm the docker harverster is started with 0 attached plots, when any of mounted subdirectories is just empty (not even an error one)

**Comment by Starttoaster** [comment]
Without more information, this Issue seems like it would belong to the chia-blockchain repository. The `recursive_plot_scan` feature is made available through code in the chia-blockchain repository and is only supported in this repository by setting it in your Chia config.yaml if you configure the environment variable.

**Comment by Starttoaster** [comment]
This isn't the behavior of the recursive plot scan feature that I'm seeing in the container either. Using this docker-compose.yml file:

```
version: "3.6"
services:
  chia:
    container_name: chia
    restart: unless-stopped
    image: ghcr.io/chia-network/chia:latest
    environment:
      plots_dir: /mnt
      keys: /key.txt
      recursive_plot_scan: "true"
    volumes:
      - /mnt:/mnt
      - /home/bb/key.txt:/key.txt
 ```

My `/mnt` directory contains several subdirectories with plots in them, and one empty subdirectory named `tempplots`. These are all mounted in the container, and on `chia farm summary` I see `Local Harvester 633 plots of size: 62.657 TiB`. 

Running `du -h /mnt` on my host, I see that the total contents come out to `63TB`, so it adds up on my end.

**Comment by Starttoaster** [comment]
Here's some output to show my directory tree:

```
bb@chia1:~$ tree /mnt
/mnt
├── final1
├── final10
│   ├── plot-k32-...plot (and more listed below this)
├── final12
├── final2
│   ├── plot-k32-...plot (and more listed below this)
├── final3
├── final4
├── final5
│   ├── lost+found
│   ├── plot-k32-...plot (and more listed below this)
├── final6
│   ├── plot-k32-...plot (and more listed below this)
├── final7
│   ├── plot-k32-...plot (and more listed below this)
├── final9
│   ├── plot-k32-...plot (and more listed below this)
└── tempplots

16 directories, 633 files
```

There are other directories that are also empty. I lost a few drives over time (we had a few heat waves in my neck of the woods that helped to fry a couple.)

**Comment by chilio** [comment]
@Starttoaster can you confirm the same if any of the directories are not readable due to for example faulty hdd?

**Comment by Starttoaster** [comment]
I can. I mentioned earlier that I've lost a few drives. I guess I never mentioned that the failed drives are still hooked up though.

Running an lsblk on my server, I see this unmounted partition. This 10TB (9.1T according to lsblk) drive used to store plots. But has since gone bad for one reason or another, I have yet to run testdisk to see if I can recover the plots off of it. 
```
sdb                         8:16   0   9.1T  0 disk
└─sdb1                      8:17   0   9.1T  0 part
```
Attempting to mount that drive returns this error:
```
bb@chia1:~$ sudo mount /dev/sdb1 /mnt/final1
mount: /mnt/final1: can't read superblock on /dev/sdb1.
```
/mnt/final1 used to be one of my plot directories but is now just an empty directory after the drive failed.

Then I start up my Chia farm with a docker-compose file looking like the following
```
version: "3.6"
services:
  chia:
    container_name: chia
    restart: unless-stopped
    image: ghcr.io/chia-network/chia:latest
    environment:
      # generic config
      TZ: "America/Vancouver"
      log_level: "INFO"
      service: "farmer"

      # plots config
      plots_dir: /mnt
      recursive_plot_scan: "true"

      # key config
      keys: "persistent"
    volumes:
      # plots
      - /mnt:/mnt
      # chia root (db, config, cache files, etc)
      - /home/bb/.chia/mainnet:/root/.chia/mainnet
      # chia keys
      - /home/bb/.chia_keys:/root/.chia_keys
```
The container starts up fine. Execing into the container I can check my farm summary to see if my harvester found its plots:
```
bb@chia1:~$ docker exec -it chia /bin/bash
root@1daf0bd5317b:/chia-blockchain# . ./activate
(venv) root@1daf0bd5317b:/chia-blockchain# chia farm summary
Farming status: Farming
Total chia farmed: 28.000000000019
User transaction fees: 1.9e-11
Block rewards: 28.0
Last height farmed: 3376023
Local Harvester
   633 plots of size: 62.657 TiB
Plot count for all harvesters: 633
Total size of plots: 62.657 TiB
Estimated network space: 21.925 EiB
Expected time to win: 2 months and 1 week
Note: log into your key using 'chia wallet show' to see rewards for each key
```
Success :) That's the expected plot count for my farm.

In summary, I also have failed hard drives, and am using the recursive plot scanning feature, and my harvester in the container is able to resolve the plots that still exist how I'd expect it to.

Going back, I asked for more details on your particular setup, because without more information, this seems potentially like an edge case that may only happen under particular circumstances, or just specific to your farm.

Alternatively... You may also want to attempt to disconnect your failed hard drives and recycle them! :)

**Comment by Starttoaster** [comment]
Worth mentioning also that this particular drive that I showed is unique in that it still shows up in lsblk even though apparently the filesystem on it is unmountable. I have another drive or two attached that are completely failed, and don't show up in lsblk, but I can verify do physically exist and are physically connected to the server, and those aren't stopping my harvester from quantifying my plots either. Figured it'd also be worth mentioning that I also have _completely_ dead plot drives, though I don't know how that should affect chia since an unmounted drive is just an empty directory.

**Comment by chilio** [comment]
@Starttoaster the problem still persists.
I don't know how you can mitigate this behavior, but it applies to using NTFS volume on linux with error file.
Sometimes there are faulty files emerging on NTFS while using linux, and you cannot do anything with them (chown, chmod, mv, absolutely nothing).
This looks like that...
![image](https://user-images.githubusercontent.com/5613137/235260706-03a9ed48-49a2-4b84-b981-373efaa4506c.png)
Actually this mentioned drive is not faulty, smart info is ok, all files are ok and I can normally use this drive.
The only problem is with this one file which causes IO errors (cannot stat) while accessing, deleting or whatsoever. 
I have not been able to find any solution how to fix it on linux.
All net solutions finally go to use chkdsk on windows which is no go in my case. I've tried this and yes windows chkdsk actually fixes these errors.
So therefore I'm still asking if there is any way that this docker can skip these errors and proceed with correct files on this drive?
Currently this chia docker image is not reading files not only from the drive where this error file is located, but also from other completely fine hdds resulting in 0 plots for this harvester.
Kindly please comment...

**Comment by Starttoaster** [comment]
Thanks for the added context to your local environment! So it sounds like the issue happens because of a lack of stability of NTFS filesystems on Linux. I would recommend moving your plots over to drives with a filesystem that is better supported by the operating system you're running Chia on, as NTFS is typically a Windows thing. This could be done with one extra drive in a rotation, where you format it as ext4 (or your favorite common Linux filesystem type) copy your plots over to that drive, delete the plots from the original drive and format it to ext4 (or similar) and continue on until you aren't mounting NTFS drives on a Linux machine.

**Comment by chilio** [comment]
@Starttoaster yeah this is my case,

I am using NTFS drives under Linux, and for sure I do not want to copy/move them to ext4 or any other filesystem.

I expect this docker image just to work, meaning => accepting all plots (from working drives, skipping eventual errored ones), and farm...

I consider this issue as a bug, if there is no solution.

Small recap...
Here is what I've observed so far:

- in my case the first mounted dir (hdd drive) seems empty, but in fact it was errored drive which was still mounted, but showing no files

- as soon as I unmounted this faulty hdd and restarted docker with the same parameters everything went well and farmer could see all the plots from this harvester and other hdds mounted under.

Therefore I believe there is a not optimized check somewhere -> if the first dir containing plots fails then the whole container fails, but it should not fail since other dirs are having legitimate plots, as in my case.

**Comment by Starttoaster** [comment]
I definitely understand it's not convenient to move plots around to different filesystems. However, we're not really discussing a bug with Chia, so much as a bug (or just a basic lack of stability) with the underlying support for NTFS Windows-native filesystems in Linux, since you're having issues inspecting your plots with basic linux commands like `ls`.

**Comment by chilio** [comment]
Hey @Starttoaster of course it is not about chia stability and general functionality. 

But we are talking about chia-docker flexibility/elasticity here, which should cover edge cases like mine for example. 

In our chia family I suppose we should solve these problems and move forward... 
For me it is not a problem -> already solved, as stated before -> manually found problematic drives.
On the other hand I believe new users might have problems with it (undocumented, coming out of the blue).
Please keep that in mind...

**Comment by Starttoaster** [comment]
Hi @chilio, can you suggest a method for better supporting NTFS in linux that chia-docker could accomplish? This seems maybe like an issue for the maintainers of libfuse or ntfs-3g to improve stability of NTFS volumes on linux.

**Comment by chilio** [comment]
Hi @Starttoaster sorry for late reply. 
To summarize this up:
1. Yes using NTFS on linux comes with certain disadvantages, but in my case just copying/moving hundreds of hdds to new filesystems requires a lot of time and effort.
2.  Regarding my docker problems, please consider these 2 cases I am facing:
 - when I mount all these drives under root host system, chia is working with 1 harvester and when any hdd drive errors or got disconnected then chia misses that drives and simply moves further on harvesting (only plots from this errored drive are missing)
 - on the other hand, when I split all these drives into slices and use them accordingly with docker harversters, then in the beginning all is fine, but as soon any of hdds mounted on harvester errors, than this harvester becomes stale, meaning having 0 plots, although it has bunch of other hdds connected to it and still working.
 - and when restarting this particular docker container with these sometimes unmounted or errored dirs, this harvester also report zero plots.
And I suppose it shuldn't work like that.
